x-pack/agent/pkg/agent/transpiler/steps.go

```diff
@@ -0,0 +1,185 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package transpiler
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+)
+
+// StepList is a container that allow the same tree to be executed on multiple defined Step.
+type StepList struct {
+	Steps []Step
+}
+
+// Step is an execution step which needs to be run.
+type Step interface {
+	Execute(rootDir string) error
+}
+
+// Execute executes a list of steps.
+func (r *StepList) Execute(rootDir string) error {
+	var err error
+	for _, step := range r.Steps {
+		err = step.Execute(rootDir)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// MarshalYAML marsharl a steps list to YAML.
+func (r *StepList) MarshalYAML() (interface{}, error) {
+	doc := make([]map[string]Step, 0, len(r.Steps))
+
+	for _, step := range r.Steps {
+		var name string
+		switch step.(type) {
+		case *DeleteFileStep:
+			name = "delete_file"
+		case *MoveFileStep:
+			name = "move_file"
+
+		default:
+			return nil, fmt.Errorf("unknown rule of type %T", step)
+		}
+
+		subdoc := map[string]Step{
+			name: step,
+		}
+
+		doc = append(doc, subdoc)
+	}
+	return doc, nil
+}
+
+// UnmarshalYAML unmarshal a YAML document into a RuleList.
+func (r *StepList) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var unpackTo []map[string]interface{}
+
+	err := unmarshal(&unpackTo)
+	if err != nil {
+		return err
+	}
+
+	// NOTE(ph): this is a bit of a hack because I want to make sure
+	// the unpack strategy stay in the struct implementation and yaml
+	// doesn't have a RawMessage similar to the JSON package, so partial unpack
+	// is not possible.
```

Comment:
Yeah I this is what I hate the most about the yaml package, there is a way to work around that https://github.com/go-yaml/yaml/issues/13#issuecomment-135407098

Reply:
I think you can remove my name next to NOTE  :)